### PR TITLE
Read kerning from GlyphsApp files

### LINKED ISSUE #51: Read kerning info
This depends on:
- googlefonts/fontra/pull/1500
- #48

We need to extract:
- kerning groups: convert glyph-level kern "keys" to a UFO style group `dict[str, list[str]]`
- kerning values

We will also need to extract vertical kerning + groups. This could be done as a separate task, but it is equally urgent.

### PR DESCRIPTION
**Pull request by ollimeier**
Fixes #51

It's related to LTR kerning, only.

I added a "V" character to the test data, so we have at least some kerning (AV) for visual testing. I also added a right kerning group to n, m, and h.

### CODE REVIEW

File: src/fontra_glyphs/backend.py [hunk 1]
```diff
@@ -457,3 +458,76 @@ def fixSourceLocations(sources, smartAxisNames):
...
+        return kernID, f"public.kern1.{kernID}"
+    # else it's a simple glyph name
+    try:
+        name = gsFont.glyphForId_(kernID).name
+        return name, name
+    except Exception:
+        return None, None
+
+
+def getNormalizedKeringDict(gsFont, gsMasterID, valueDicts):
```
Comment (justvanrossum) [inline src/fontra_glyphs/backend.py:475]: `getNormalizedKeringDict` -> `getNormalizedKerningDict`

File: src/fontra_glyphs/backend.py [hunk 2]
```diff
@@ -457,3 +458,76 @@ def fixSourceLocations(sources, smartAxisNames):
         for source in sources:
             if source.location.get(axis) == value:
                 del source.location[axis]
+
+
+def getKerningNameFormID(gsFont, kernID):
```
Comment (justvanrossum) [inline src/fontra_glyphs/backend.py:463]: `getKerningNameFormID` -> `getKerningNameFromID`?

**Review by justvanrossum — COMMENTED** [review]
Apart from two typos: you will have to extract the kern keys from `self.rawGlyphsData` via `self.glyphNameToIndex`. The backend loads/parses the glyphs lazily, so they are not guaranteed to exist in `self.gsFont.glyphs`. We want to avoid loading all glyphs when reading kerning.

File: src/fontra_glyphs/backend.py (hunk 1 above)
Comment (ollimeier) [reply src/fontra_glyphs/backend.py]: thanks.done.

**Comment by ollimeier** [comment]
> Apart from two typos: you will have to extract the kern keys from `self.rawGlyphsData` via `self.glyphNameToIndex`. The backend loads/parses the glyphs lazily, so they are not guaranteed to exist in `self.gsFont.glyphs`. We want to avoid loading all glyphs when reading kerning.

May I ask you to get more information about 'rawGlyphsData'? I don't know what this is about and how to use it. Or maybe I simply don't understand what to do.

**Comment by justvanrossum** [comment]
> May I ask you to get more information about 'rawGlyphsData'? I don't know what this is about and how to use it. Or maybe I simply don't understand what to do.

It's in the code you are editing, backend.py, please try to read the code.

Parsing glyphs files is a two-step process:
1. parse the (ascii) plist format into raw Python data structures (bool/int/float/string, list, dict, just like json.load would do)
2. construct higher level objects as defined and implemented by glyphsLib

Step 1 we cannot avoid (and is very fast, thanks to the openstep-plist module), but step 2 is done lazily, on-demand. In backend.py, this is implemented by the `_ensureGlyphIsParsed()` method.

The "raw" glyph data is available as the `self.rawGlyphsData` list, and you index it via the `self.glyphNameToIndex` mapping. Please have a look at that data. It corresponds _directly_ with the raw .glyphs/.glyphspackage glyph data on disk.

**Comment by ollimeier** [comment]
I am closing this PR, because we will merge it into #48

**Pull request closed without merging**